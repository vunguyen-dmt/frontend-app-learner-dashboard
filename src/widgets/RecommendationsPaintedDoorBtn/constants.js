export const COLLAPSED_NAVBAR = 'collapsedNavbar';
export const EXPANDED_NAVBAR = 'expendedNavbar';
export const RECOMMENDATIONS_PANEL = 'recommendationsPanel';
