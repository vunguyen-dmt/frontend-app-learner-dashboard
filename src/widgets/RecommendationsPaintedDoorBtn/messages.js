import { defineMessages } from '@edx/frontend-platform/i18n';

const messages = defineMessages({
  recommendationsFeatureText: {
    id: 'RecommendationsPanel.recommendationsFeatureText',
    defaultMessage: 'Personalized recommendations feature is not yet available. We are working hard on bringing it to your learner home in the near future.',
    description: 'recommendations feature text',
  },
  recommendationsAlertText: {
    id: 'RecommendationsPanel.recommendationsAlertText',
    defaultMessage: 'Would you like to be alerted when it becomes available?',
    description: 'recommendations alerted text',
  },
  recommendationsModalHeading: {
    id: 'RecommendationsPanel.recommendationsModalHeading',
    defaultMessage: 'Thank you for your interest!',
    description: 'Heading of modal',
  },
  modalSkipButton: {
    id: 'RecommendationsPanel.modalSkipButton',
    defaultMessage: 'Skip for now',
    description: 'button for Skip for now',
  },
  modalCountMeButton: {
    id: 'RecommendationsPanel.modalCountMeButton',
    defaultMessage: 'Count me in!',
    description: 'button for Count me in!',
  },
  recommendedForYou: {
    id: 'learnerVariantDashboard.recommendedForYou',
    defaultMessage: 'Recommended For You',
    description: 'Header link for recommended page.',
  },
  seeAllRecommendationsButton: {
    id: 'RecommendationsPanel.seeAllRecommendationsButton',
    defaultMessage: 'See All Recommendations',
    description: 'Button to see all recommendations',
  },
});

export default messages;
