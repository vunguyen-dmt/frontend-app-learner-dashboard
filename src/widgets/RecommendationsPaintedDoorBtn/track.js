import { StrictDict } from 'utils';
import { createEventTracker } from 'data/services/segment/utils';

export const eventNames = StrictDict({
  variationGroup: 'edx.bi.user.recommendation_home.variation.group',
  recommendationHomeBtnClicked: 'edx.bi.user.recommendation_home.btn.clicked',
  recommendationHomeModalInterestBtnClicked: 'edx.bi.user.recommendation_home.modal.interest_btn.clicked',
  recommendationHomeModalSkipBtnClicked: 'edx.bi.user.recommendation_home.modal.skip_btn.clicked',
});

export const trackPaintedDoorVariationGroup = (variation) => {
  createEventTracker(eventNames.variationGroup, {
    variation,
    page: 'dashboard',
  })();
};

export const trackPaintedDoorRecommendationHomeBtnClicked = (variation) => {
  createEventTracker(eventNames.recommendationHomeBtnClicked, {
    variation,
    page: 'dashboard',
  })();
};

export const trackPaintedDoorRecommendationHomeInterestBtnClicked = (variation) => {
  createEventTracker(eventNames.recommendationHomeModalInterestBtnClicked, {
    variation,
  })();
};

export const trackPaintedDoorRecommendationHomeSkipBtnClicked = (variation) => {
  createEventTracker(eventNames.recommendationHomeModalSkipBtnClicked, {
    variation,
  })();
};
