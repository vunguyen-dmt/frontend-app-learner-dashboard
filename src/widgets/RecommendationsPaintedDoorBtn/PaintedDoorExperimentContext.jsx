import React from 'react';
import PropTypes from 'prop-types';
import { StrictDict } from 'utils';
import * as module from './PaintedDoorExperimentContext';
import {
  useEmailConfirmationData,
  useHasAvailableDashboards,
  useRequestIsPending,
} from '../../data/redux/hooks';
import { RequestKeys } from '../../data/constants/requests';
import { trackPaintedDoorVariationGroup } from './track';

export const state = StrictDict({
  enterpriseUser: (val) => React.useState(val), // eslint-disable-line
  experimentData: (val) => React.useState(val), // eslint-disable-line
});

const PAINTED_DOOR_RECOMMENDATIONS_EXP_ID = 25116810832;
const PAINTED_DOOR_RECOMMENDATIONS_PAGE = 'url_targeting_for_van1604_recommendations_painted_door_exp';
const PAINTED_DOOR_RECS_EXP_NAVBAR_BTN_VARIATION = 'btn_navbar';
const PAINTED_DOOR_RECS_EXP_WIDGET_BTN_VARIATION = 'btn_widget';
const PAINTED_DOOR_RECS_EXP_CONTROL_VARIATION = 'control';

export function getPaintedDoorRecommendationsExperimentVariation() {
  try {
    if (window.optimizely && window.optimizely.get('data').experiments[PAINTED_DOOR_RECOMMENDATIONS_EXP_ID]) {
      const selectedVariant = window.optimizely.get('state').getVariationMap()[PAINTED_DOOR_RECOMMENDATIONS_EXP_ID];
      return selectedVariant?.name;
    }
  } catch (e) { /* empty */ }
  return '';
}

export function activatePaintedDoorRecommendationsExperiment() {
  window.optimizely = window.optimizely || [];
  window.optimizely.push({
    type: 'page',
    pageName: PAINTED_DOOR_RECOMMENDATIONS_PAGE,
  });
}

export const useIsEnterpriseUser = () => {
  const [enterpriseUser, setEnterpriseUser] = module.state.enterpriseUser({
    isEnterpriseUser: false,
    isLoading: true,
  });

  const initIsPending = useRequestIsPending(RequestKeys.initialize);
  const hasAvailableDashboards = useHasAvailableDashboards();
  const confirmationData = useEmailConfirmationData();

  React.useEffect(() => {
    if (!initIsPending && Object.keys(confirmationData).length && hasAvailableDashboards) {
      setEnterpriseUser(prev => ({
        ...prev,
        isEnterpriseUser: true,
        isLoading: false,
      }));
    } else if (!initIsPending && Object.keys(confirmationData).length && !hasAvailableDashboards) {
      setEnterpriseUser(prev => ({
        ...prev,
        isEnterpriseUser: false,
        isLoading: false,
      }));
    }
  }, [initIsPending]); // eslint-disable-line react-hooks/exhaustive-deps

  return enterpriseUser;
};

export const PaintedDoorExperimentContext = React.createContext({
  experimentVariation: null,
  isPaintedDoorNavbarBtnVariation: null,
  isPaintedDoorWidgetBtnVariation: null,
  isPaintedDoorControlVariation: null,
  experimentLoading: null,
});

export const PaintedDoorExperimentProvider = ({ children }) => {
  const [experimentData, setExperimentData] = module.state.experimentData({
    experimentVariation: '',
    isPaintedDoorNavbarBtnVariation: false,
    isPaintedDoorWidgetBtnVariation: false,
    isPaintedDoorControlVariation: false,
    experimentLoading: true,
  });
  const enterpriseUser = useIsEnterpriseUser();

  const contextValue = React.useMemo(
    () => ({
      ...experimentData,
    }),
    [experimentData],
  );

  React.useEffect(() => {
    let timer = null;
    if (!enterpriseUser.isLoading && !enterpriseUser.isEnterpriseUser) {
      activatePaintedDoorRecommendationsExperiment();
      timer = setTimeout(() => {
        const variation = getPaintedDoorRecommendationsExperimentVariation();
        setExperimentData(prev => ({
          ...prev,
          experimentVariation: variation,
          isPaintedDoorNavbarBtnVariation: variation === PAINTED_DOOR_RECS_EXP_NAVBAR_BTN_VARIATION,
          isPaintedDoorWidgetBtnVariation: variation === PAINTED_DOOR_RECS_EXP_WIDGET_BTN_VARIATION,
          isPaintedDoorControlVariation: variation === PAINTED_DOOR_RECS_EXP_CONTROL_VARIATION,
          experimentLoading: false,
        }));
        trackPaintedDoorVariationGroup(variation);
      }, 500);
    }
    return () => clearTimeout(timer);
  }, [enterpriseUser]); // eslint-disable-line react-hooks/exhaustive-deps

  return (
    <PaintedDoorExperimentContext.Provider value={contextValue}>
      {children}
    </PaintedDoorExperimentContext.Provider>
  );
};

export const usePaintedDoorExperimentContext = () => React.useContext(PaintedDoorExperimentContext);

PaintedDoorExperimentProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

export default PaintedDoorExperimentProvider;
