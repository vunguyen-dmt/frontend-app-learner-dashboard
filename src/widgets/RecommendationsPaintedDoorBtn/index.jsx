import React from 'react';
import PropTypes from 'prop-types';
import NavbarButton from './components/NavbarButton';
import PaintedDoorModal from './components/PaintedDoorModal';
import { usePaintedDoorModal } from './components/PaintedDoorModal/hooks';
import { COLLAPSED_NAVBAR, EXPANDED_NAVBAR, RECOMMENDATIONS_PANEL } from './constants';
import RecommendationsPanelButton from './components/RecommendationsPanelButton';
import { trackPaintedDoorRecommendationHomeBtnClicked } from './track';

export const RecommendationsPaintedDoorBtn = ({ placement, experimentVariation }) => {
  const { isModalOpen, toggleModal } = usePaintedDoorModal();

  const handleClick = () => {
    toggleModal();
    trackPaintedDoorRecommendationHomeBtnClicked(experimentVariation);
  };

  const getPaintedDoorButton = () => {
    if ([COLLAPSED_NAVBAR, EXPANDED_NAVBAR].includes(placement)) {
      return (
        <NavbarButton handleClick={handleClick} placement={placement} />
      );
    } if (placement === RECOMMENDATIONS_PANEL) {
      return (
        <RecommendationsPanelButton handleClick={handleClick} />
      );
    }
    return null;
  };

  return (
    <>
      {getPaintedDoorButton()}
      <PaintedDoorModal isOpen={isModalOpen} onClose={toggleModal} variation={experimentVariation} />
    </>
  );
};

RecommendationsPaintedDoorBtn.propTypes = {
  placement: PropTypes.oneOf([COLLAPSED_NAVBAR, EXPANDED_NAVBAR, RECOMMENDATIONS_PANEL]).isRequired,
  experimentVariation: PropTypes.string.isRequired,
};

export default RecommendationsPaintedDoorBtn;
