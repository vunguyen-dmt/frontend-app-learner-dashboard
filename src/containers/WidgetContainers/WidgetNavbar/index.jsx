import React from 'react';
import PropTypes from 'prop-types';
import RecommendationsPaintedDoorBtn from 'widgets/RecommendationsPaintedDoorBtn';
import { COLLAPSED_NAVBAR, EXPANDED_NAVBAR } from 'widgets/RecommendationsPaintedDoorBtn/constants';
import {
  usePaintedDoorExperimentContext,
} from 'widgets/RecommendationsPaintedDoorBtn/PaintedDoorExperimentContext';

export const WidgetNavbar = ({ placement }) => {
  const {
    experimentVariation,
    isPaintedDoorNavbarBtnVariation,
    experimentLoading,
  } = usePaintedDoorExperimentContext();

  if (!experimentLoading && isPaintedDoorNavbarBtnVariation) {
    return (
      <RecommendationsPaintedDoorBtn placement={placement} experimentVariation={experimentVariation} />
    );
  }

  return null;
};

WidgetNavbar.propTypes = {
  placement: PropTypes.oneOf([COLLAPSED_NAVBAR, EXPANDED_NAVBAR]).isRequired,
};

export default WidgetNavbar;
