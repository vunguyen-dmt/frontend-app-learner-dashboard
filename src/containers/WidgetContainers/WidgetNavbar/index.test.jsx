import { shallow } from 'enzyme';
import {
  usePaintedDoorExperimentContext,
} from '../../../widgets/RecommendationsPaintedDoorBtn/PaintedDoorExperimentContext';
import WidgetNavbar from './index';
import { EXPANDED_NAVBAR } from '../../../widgets/RecommendationsPaintedDoorBtn/constants';
import RecommendationsPaintedDoorBtn from '../../../widgets/RecommendationsPaintedDoorBtn';

jest.mock('widgets/RecommendationsPaintedDoorBtn/PaintedDoorExperimentContext', () => ({
  usePaintedDoorExperimentContext: jest.fn(),
}));

describe('WidgetNavbar', () => {
  let mockExperimentContext = {
    experimentVariation: '',
    isPaintedDoorNavbarBtnVariation: true,
    experimentLoading: false,
  };
  const props = {
    placement: EXPANDED_NAVBAR,
  };

  describe('snapshots', () => {
    test('default', () => {
      usePaintedDoorExperimentContext.mockReturnValueOnce(mockExperimentContext);
      const wrapper = shallow(<WidgetNavbar {...props} />);

      expect(usePaintedDoorExperimentContext).toHaveBeenCalled();
      expect(wrapper).toMatchSnapshot();
    });
  });

  test('renders button if user in navbar variation', () => {
    usePaintedDoorExperimentContext.mockReturnValueOnce(mockExperimentContext);
    const wrapper = shallow(<WidgetNavbar {...props} />);

    expect(usePaintedDoorExperimentContext).toHaveBeenCalled();
    expect(wrapper.type()).toBe(RecommendationsPaintedDoorBtn);
  });

  test('renders nothing if user in not in navbar variation', () => {
    mockExperimentContext = {
      ...mockExperimentContext,
      isPaintedDoorNavbarBtnVariation: false,
    };
    usePaintedDoorExperimentContext.mockReturnValueOnce(mockExperimentContext);
    const wrapper = shallow(<WidgetNavbar {...props} />);

    expect(usePaintedDoorExperimentContext).toHaveBeenCalled();
    expect(wrapper.type()).toBeNull();
  });

  test('renders nothing if experiment is loading', () => {
    mockExperimentContext = {
      ...mockExperimentContext,
      isPaintedDoorNavbarBtnVariation: false,
      experimentLoading: true,
    };
    usePaintedDoorExperimentContext.mockReturnValueOnce(mockExperimentContext);
    const wrapper = shallow(<WidgetNavbar {...props} />);

    expect(usePaintedDoorExperimentContext).toHaveBeenCalled();
    expect(wrapper.type()).toBeNull();
  });
});
