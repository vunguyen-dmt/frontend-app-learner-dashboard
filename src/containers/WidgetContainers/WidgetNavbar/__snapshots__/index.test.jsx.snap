// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WidgetNavbar snapshots default 1`] = `
<RecommendationsPaintedDoorBtn
  experimentVariation=""
  placement="expendedNavbar"
/>
`;
